src/node/nodestate.h

```diff
@@ -786,35 +789,78 @@ namespace ccf
     }
 
     bool finish_recovery(
-      Store::Tx& tx, const nlohmann::json& sealed_secrets) override
+      Store::Tx& tx,
+      const nlohmann::json& sealed_secrets,
+      bool with_shares) override
     {
       std::lock_guard<SpinLock> guard(lock);
       sm.expect(State::partOfPublicNetwork);
 
-      LOG_INFO_FMT("Initiating end of recovery (primary)");
+      if (with_shares)
```

Comment:
If we are doing a recovery with shares, we only change the state of the service and wait for members to submit their shares. `finish_recovery_with_shares()` below will be called once members have submitted their shares.